source/extensions/filters/network/thrift_proxy/compact_protocol_impl.cc

```diff
@@ -111,14 +112,20 @@ bool CompactProtocolImpl::peekReplyPayload(Buffer::Instance& buffer, ReplyType&
     return false;
   }
 
-  if (id < 0 || id > std::numeric_limits<int16_t>::max()) {
-    throw EnvoyException(absl::StrCat("invalid compact protocol field id ", id));
-  }
+  validateFieldId(id);
   // successful response struct in field id 0, error (IDL exception) in field id greater than 0
   reply_type = id == 0 ? ReplyType::Success : ReplyType::Error;
   return true;
 }
 
+void CompactProtocolImpl::validateFieldId(int32_t id) {
+  if ((id < 0 && !Runtime::runtimeFeatureEnabled(
+                     "envoy.reloadable_features.thrift_allow_negative_field_ids")) ||
+      id < std::numeric_limits<int16_t>::min() || id > std::numeric_limits<int16_t>::max()) {
+    throw EnvoyException(absl::StrCat("invalid compact protocol field id ", id));
```

Comment:
nit: this is pretty hard to read -- do an initial check for ids in the always-valid range and return early and then do the extra check for the flag to decide whether to throw or not?

Reply:
I refactored both protocol's check to the suggested style.